### PR TITLE
adapter: don't optimize read transactions for single statements

### PR DESCRIPTION
**Pull request by mjibson**
This fixes the csharp test flakes. See parse-started.pt for the full explanation.

### Motivation

  * This PR fixes a previously unreported bug.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-protobuf` label.
- [X] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - When `mz_logical_timestamp()` is called within a view, it will ALWAYS return the current timestamp of it's timedomain. The old behavior was to return `Timestamp::MAX` outside of a transaction if no other views depended on it.
  - Queries in the `Serializable` isolation level outside of a transaction may not return the freshest possible data for the object being queried. Instead it will use a timestamp that is immediately query-able over all objects in the same time domain.

### CODE REVIEW

**Comment by mjibson** [comment]
Ugh this is harder than I thought. Of course tons of stuff does single statement DDL over extended protocol, so this solution won't work. I think we have to enforce this at the coord layer or something? It's verrry tricky to support single statement DDL *and* support Npgsql.

**Comment by mjibson** [comment]
RFAL. This doesn't fix the underlying problem, but does stop the panic and now clearly documents what we are doing wrong instead of having it be accidental behavior. I'm not sure what the real fix would be. Maybe something like have DDL statements register a transaction op that's similar to TAIL's one statement only thing? In any case this should unblock CI.

**Comment by jkosh44** [comment]
One of the reasons that this isn't passing CI is because of this issue: https://github.com/MaterializeInc/materialize/issues/14024

**Comment by mjibson** [comment]
Which thing is reading from an introspection source?

**Comment by jkosh44** [comment]
> Which thing is reading from an introspection source?

cluster_log_sinks.slt is failing and reading from an introspection source. I have a fix incoming, but I'll probably open a separate PR.

**Comment by jkosh44** [comment]
Another issue is that `SELECT mz_logical_timestamp()` always returns `18446744073709551615`. I think this has something to do with Implicit read transactions not initializing all the correct transaction state and the transaction timestamp accidentally gets initialized to `TIMESTAMP::MAX`. The issue manifests itself in `sequence_peek` at `dataflow.set_as_of(Antichain::from_elem(timestamp));`.

I'll try and continue to debug tonight or tomorrow.

File: src/adapter/src/coord/sequencer.rs
Comment (mjibson) [inline src/adapter/src/coord/sequencer.rs:2407]: Why did this change need to happen? Adding all these IDs to the explain timestamp output seems bad. Won't this make something like `explain timestamp for select * from t` also print all other objects in t's schema?

File: test/testdrive/timestamps-debezium-kafka.td
Comment (mjibson) [inline test/testdrive/timestamps-debezium-kafka.td:192]: What happened here?

File: test/testdrive/numeric.td
Comment (mjibson) [inline test/testdrive/numeric.td:40]: What caused these changes?

File: test/testdrive/csv-sources.td
Comment (mjibson) [inline test/testdrive/csv-sources.td:140]: :( this kinda makes explain timestamp bad now

File: src/adapter/src/coord/sequencer.rs
Comment (jkosh44) [reply src/adapter/src/coord/sequencer.rs:2407]: Yes, but that's how timestamps are determined now so it'd be inaccurate to leave them out. When choosing a timestamp for a query we use all IDs in the time-domain in all isolation levels.

File: test/testdrive/timestamps-debezium-kafka.td
Comment (jkosh44) [reply test/testdrive/timestamps-debezium-kafka.td:192]: You can no longer view the most recent data in `bar` until `foo` catches up since they are in the same time domain.

File: test/testdrive/numeric.td
Comment (jkosh44) [reply test/testdrive/numeric.td:40]: Previously `mz_logical_timestamp()` would return `Timestamp::MAX` in a view which made this true. Now it returns the current timestamp of the timeline it's in making it false.

File: src/adapter/src/coord/command_handler.rs
Comment (mjibson) [reply src/adapter/src/coord/command_handler.rs:290]: https://github.com/MaterializeInc/materialize/issues/14038

File: src/adapter/src/coord/sequencer.rs
Comment (mjibson) [reply src/adapter/src/coord/sequencer.rs:2407]: Ah so if we did leave this as is, then it's just misleading? Hmm can we teach explain timestamp to special case all the other IDs into a single thing that's timeline information or something? I really think we should keep the output of this trimmed down to print things referenced by the query.

Comment (mjibson) [reply src/adapter/src/coord/sequencer.rs:2407]: I'd almost prefer leaving this as is and fixing it in another pr.

Comment (jkosh44) [reply src/adapter/src/coord/sequencer.rs:2407]: I think misleading isn't strong enough. Without this change the output is wrong for serializable. Image you have two objects in the same time domain with the following `[since, upper)` ranges: A: `[10, 20)` and B: `[13, 25)`. The timestamp selected to read B under Serializable will be `19`. However, the old explain timestamp will show `24`.

What we could do is add two cases, one for Strict Serializable and one for Serializable. The Strict Serializable one doesn't pull in the whole time domain while the Serializable one does pull in the whole time domain. That way it's still clean for Strict Serializable but also accurate for Serializable.

**Review by jkosh44 — APPROVED** [review]
LGTM

**Comment by jkosh44** [comment]
@mjibson I can't officially add you as a reviewer since you're the author, but I'll hold off on merging until you sign off. Let me know your thoughts on the new EXPLAIN stuff that I pushed.

File: src/environmentd/tests/sql.rs
Comment (mjibson) [inline src/environmentd/tests/sql.rs:1384]: Does this no longer pass?

File: src/environmentd/tests/sql.rs
Comment (mjibson) [inline src/environmentd/tests/sql.rs:1546]: would `AS OF AT LEAST` help here?

File: test/testdrive/explain-timestamps.td
Comment (mjibson) [inline test/testdrive/explain-timestamps.td:16]: Am I correct that this comment is only true from EXPLAIN's point of view? sequence_peek doesn't about time domain, and neither does determine_timestamp.

**Review by mjibson — COMMENTED** [review]
LGTM

File: src/environmentd/tests/sql.rs
Comment (jkosh44) [reply src/environmentd/tests/sql.rs:1546]: Yes! Thanks for the idea.

File: test/testdrive/explain-timestamps.td
Comment (jkosh44) [reply test/testdrive/explain-timestamps.td:16]: I think this is half true from SELECT's point of view too. Every ID in the timedomain is passed into `determine_timestamp`, but in Strict Serializable mode we don't look at them except to determine the timeline.

File: src/environmentd/tests/sql.rs
Comment (jkosh44) [reply src/environmentd/tests/sql.rs:1384]: This no longer passes since the table exists at the time of the view query it will push the timestamp of the view query back to the `upper - 1` of the table. Then the join will also be at `upper - 1` of the table.

In the above serializable assert, we make sure to query the view before the table exists. So the timestamp of the view query is `upper - 1` of the view.

**Pull request merged**
